proposed/resource-limit.md

```diff
@@ -0,0 +1,843 @@
+{
+    // an inaccurate view of resources
+    public abstract long EstimatedCount(TKey resourceID);
+
+    // Fast synchronous attempt to acquire resources
+    public abstract Resource Acquire(TKey resourceID, long requestedCount);
+
+    // Wait until the requested resources are available
+    // If unsuccessful, throw
+    public abstract ValueTask<Resource> AcquireAsync(TKey resourceID, long requestedCount, CancellationToken cancellationToken = default);
+}
+```
+
+See a [proof of concept implementation](#ip-address-aggregated-resource-limiter-using-a-local-concurrentdictionary-as-a-storage-mechanism).
+
+## Stakeholders and Reviewers
+
+- Owned by ASP.NET Core
+- Code will be a new area in dotnet/runtime
+  - Tentatively System.Threading.ResourceLimits
+- Reviewers
+  - David Fowler, Eric Erhardt, Stephen Halter, Chris Ross
+
+## Design
+
+### Abstractions
+
+The main abstraction API will consist of the `ResourceLimiter` and the associated `Resource`. The main responsibility of the `ResourceLimiter` is to allow the user to determine whether it's possible to proceed with an operation. The `Resource` struct is returned by the `ResourceLimiter` to represent if the acquisition was successful and the resources that were obtained, if successful.
+
+```c#
+    // Represents a limiter type that users interact with to determine if an operation can proceed
+    public abstract class ResourceLimiter
+    {
+        // An estimated count of the underlying resources
+        abstract long EstimatedCount { get; }
+
+        // Fast synchronous attempt to acquire resources
+        // Set requestedCount to 0 to get whether resource limit has been reached
+        abstract Resource Acquire(long requestedCount);
+
+        // Wait until the requested resources are available
+        // Set requestedCount to 0 to wait until resource is replenished
+        abstract ValueTask<Resource> AcquireAsync(long requestedCount, CancellationToken cancellationToken = default);
+    }
+
+    // Represents a resource obtained from the limiter. The user disposes this type to release the acquired resources.
+    public struct Resource : IDisposable
+    {
+        // This represents whether resource acquisition was successful
+        public bool IsAcquired{ get; }
+
+        // This represents additional metadata that can be returned as part of a call to Acquire/AcquireAsync
+        // Potential uses could include a RetryAfter value or an error code.
+        public object? State { get; }
+
+        // `state` is the additional metadata, onDispose takes `state` as its argument.
+        public Resource(bool isAcquired, object? state, Action<Resource>? onDispose);
+
+        public void Dispose();
+
+        // This static `Resource` is used by rate limiters or in cases where `Acquire` returns false.
+        public static Resource SuccessNoopResource = new Resource(true, null, null);
+        public static Resource FailNoopResource = new Resource(false, null, null);
```

Comment:
Do we need these anymore?

Reply:
Yes I think so, these are convenient for quick default limiter responses. For example, they would be used by all rate limiters and when concurrency limiters fail. FYI I've also updated the names to `SuccessfulAcquisition` and `FailedAcquisition`. I'm still not 100% happy with the names though.

Reply:
Removed